### ISSUE TITLE
Re-export engine creation function

### ISSUE THREAD

**Issue by kajacx** [DouglasDwyer/wasm_runtime_layer#3]
I was really interested about this *backend-agnostic wasm api* idea, but I was kind of thrown back when I noticed this line in the example:

```rust
let engine = Engine::new(wasmi::Engine::default());
```

This is kind of bad, because I have to add wasmi to my dependencies, making it less "backend agnostic" then I would like. Ideally, if only one backend feature is selected, it should be possible to "just create the engine" and not care about from which backend that engine comes from.

If I then want to change the backend from wasmi to wasmtime, for example, I should be able to just replace the feature flag in my Cargo.toml and not touch anything else, and it should just use the new backend. That is the truly backend-agnostic way.

However, there are situation when that is not desired. You could, for example, want both backends at the same time, each for a different module. Activating both features and "just creating the engine" in this way would obviously not work, you would have to specify *which* engine you want.

But there is a better way. Instead of adding the backend dependency to your Cargo.toml and doing god-knows-what to hunt for the correct version for it, you could just re-export the engine creation methods. So for example, instead of `wasmi::Engine::default()` you would write `wasm_runtime_layer::wasmi::Engine::default()`. You are still "choosing" the wasmi backend, but without the need to add wasmi to your dependencies.

But returning to the "just make the engine" approach, if exactly one backend feature is active, it would be nice to be able to do just `Engine::default()` and have it work with that backend.

**Comment by kajacx** [comment]
For example, [here](https://github.com/kajacx/wasm-playground/blob/bevy-wasm-runtime/host/src/main.rs#L76-L84) I need to do this annoying "function fork" depending on target:
```rust
#[cfg(not(target_arch = "wasm32"))]
fn get_engine() -> impl backend::WasmEngine {
    wasmtime::Engine::default()
}

#[cfg(target_arch = "wasm32")]
fn get_engine() -> impl backend::WasmEngine {
    web::Engine::default()
}
```
It's not the end of the world, but it would be better if there was one single function that would work with any engine.

Also, I don't know if this is related or not, but the `sys` (non-wasm) version of the code doesn't work. I get an `the trait 'WasmEngine' is not implemented for 'wasmtime::Engine'` error message, even though I wrote the code based on [one of the test examples](https://github.com/DouglasDwyer/wasm_runtime_layer/blob/master/tests/add_one.rs#L7).

This could be becase it doesn't work with wasmtime version 16, which, again, how I am supposed to know which versions of wasmtime it works with and doesn't work with? If it simply re-exported wasmtime, I could use `wasm_runtime_layer::wasmtime` and have a 100% guarantee that I am using the correct version.

**Comment by DouglasDwyer** [comment]
Hey @kajacx, thanks for the feedback. This is helpful and I appreciate your perspective. There are a couple reasons that the API is designed the way that it is.

- First off, it's impossible to implement `Engine::default` since `Engine` is a generic type, and the user would be required to specify it. But it would be possible to create a `fn default_engine() -> impl WasmEngine` to do the same thing.
- However, this breaks the rule that features should be additive. The Cargo book states that "enabling a feature should not disable functionality, and it should usually be safe to enable any combination of features." Imagine this: you are using `wasm_runtime_layer` with the `backend_web` crate, so the default engine functionality is available. Now, one of your dependencies releases a Semver-compatible update which also pulls in `wasm_runtime_layer` and enables another backend. This causes your code to no longer compile. Overall, I don't think that the extra convenience of a default implementation is worth allowing situations like this.
- So that leaves us with re-exportation. The reason that the end-user is required to manually bring in a dependency is because I wanted them to be able to specify and control the backend crate version themselves. If you look at the project's `Cargo.toml`, the `wasmtime` and `wasmi` crates only have a minimum version specified. This gives increased flexibility to the end user. Albeit, this does come at the expense of future updates to `wasmtime` or `wasmi` breaking things, so maybe we should specify the version.

For the reasons detailed above, I don't think that I want to introduce a default implementation. I am open to adding re-exports, but I still think it's good to have the user manually specify/import the backend crate version themselves since this crate does not impose strong version requirements. What do you think?

Regarding your trait implementation example, that's very surprising. I will clone your `wasm-playground` example and take a look.

**Comment by DouglasDwyer** [comment]
I am able to build `wasm-playground/host` just fine on Windows. I would recommend checking `cargo tree -p wasmtime` and running `cargo update -p wasmtime` if there are multiple versions of it in your project.

I guess maybe the best thing to do would be to have multiple feature flags (like `wasmtime_0_16` and `wasmtime_0_17`) so that it's clear to the end user what version is being re-exported and the user still has version control. Would this solution be acceptable to you @kajacx?

**Comment by kajacx** [comment]
You are right about the additive features thing. You still can put the engine selection behind feature flags of your own, and then enable `wasm_runtime_layer`'s features based on that as well, so in the end you would still be able to switch which backend you use simply by using a different feature, without changing a single line of code. But that would make *your* code not feature-additive-proof, unless you *prioritize* which backend is used in case of multiple features enabled. But that is for the *user* of the `wasm_runtime_layer` crate to decide, and not something that should be in the crate itself.

With the "not compiling" example, I deleted my Cargo.lock and target folders, and now it runs fine. The joys of non-reproducible errors.

The "re-export" feature is the one that deserves most consideration. It depend on what is the "philosophy" of the crate. You could:

1. Make it a complete package with multiple built-in backends that the user can choose from (or add their own custom one)
2. Make it a "conversion" layer that users can put on top of the backend(s) which they are already using

In the first case, it makes sense to re-export the engine creation methods, so that the users don't have to add anything more to their Cargo.toml files. In second case, it makes more sense to already have the backend in your Cargo.toml anyway and pass that crate's engine to the "conversion" engine.

I personally like the first approach more, the more something works out of the box with minimal configuration, the better.

When it comes to the "users not having enough control over the backend crate", they can still add it to their Cargo.toml and have control over it's version and features, it's just an *optional* thing they can do if they need that more control, and not something they *have to* do, regardless of whether they need that extra control or not.

**Comment by DouglasDwyer** [comment]
I am closing this issue, because I believe that the main concerns have been addressed in #9 (or, at least, we've chosen the philosophy of the project). Each WASM backend is now its own crate. This has the following advantages:

- There will be no more issues with different backend versions or non-reproducible compilation issues. Backend crates are versioned based upon the underlying implementation (for instance, `wasmi_runtime_layer` has version `0.31.0`, which corresponds to `wasmi` version `0.31.0`). This gives the user ultimate control over which backend versions they are choosing.
- Each of the WASM backend crates re-export their runtime. So you can do `wasmi_runtime_layer::Engine::default()` without having to add `wasmi` itself to your dependencies.

Ultimately, the goal of this project is to facilitate writing libraries like my `wasm_component_layer` or `wings` crates, which work with multiple WASM runtimes. Having the user specify which backend they want (`wasmi`, `wasmtime`, etc.) seems inevitable to me for building binaries, but libraries can be built agnostically. I think that this approach will best support that goal!
